### PR TITLE
Enh/liquid motors tank inertia

### PR DESCRIPTION
**Pull request by phmbressan**
## Pull request type

Please check the type of change your PR introduces:

- [x] Code base additions (bugfix, features)
- [ ] Code maintenance (refactoring, formatting, renaming, tests)
- [ ] ReadMe, Docs and GitHub maintenance
- [ ] Other (please describe):

## Pull request checklist

- Code base additions (for bug fixes / features):

  - [ ] Tests for the changes have been added
  - [ ] Docs have been reviewed and added / updated if needed
  - [x] Lint (`black rocketpy`) has passed locally and any fixes were made
  - [ ] All tests (`pytest --runslow`) have passed locally

## What is the current behavior?

There is no support for non flat cap inertia evaluation nor the tank gaseous volume.

## What is the new behavior?

Expand inertia evaluation for Tanks. The new behavior includes spherical cap inertia calculations and the gaseous portion is taken into account.

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

**Comment by giovaniceotto** [comment]
Currently working on this review but haven't finished yet. Expect to finish it by today.

File: rocketpy/motors/Tank.py [hunk 1]
```diff
@@ -0,0 +1,842 @@
...
+        self.bottomCap = bottomCap
+        self.upperCap = upperCap
+
+        self.capMap = {
+            "flat": Disk,
+            "spherical": Hemisphere,
+        }
+        self.setGeometry()
+
+        pass
```
Comment (giovaniceotto) [inline rocketpy/motors/Tank.py:35]: Consider removing unnecessary lines.

```suggestion
```

File: rocketpy/motors/Tank.py [hunk 2]
```diff
@@ -0,0 +1,842 @@
...
+        """Sets the geometry of the tank based on the input parameters.
+
+        Returns
+        -------
+        None
+        """
+        self.cylinder = Cylinder(self.diameter / 2, self.height)
+        self.bottomCap = self.capMap.get(self.bottomCap)(
+            self.diameter / 2, fill_direction="upwards"
+        )
```
Comment (giovaniceotto) [inline rocketpy/motors/Tank.py:47]: Should we add error handling to deal with non-existent cap types?

File: rocketpy/motors/Tank.py [hunk 3]
```diff
@@ -0,0 +1,842 @@
...
+        self.upperCap = self.capMap.get(self.upperCap)(
+            self.diameter / 2, fill_direction="downwards"
+        )
+
+    def evaluateTankState(self, t):
+        """Gets the state of the tank at a given time: the volume of liquid and
+        gases at each part of the tank (body and caps) as well as its masses.
```
Comment (giovaniceotto) [inline rocketpy/motors/Tank.py:54]:
```suggestion
        gases at each part of the tank (body and caps) as well as their masses.
```

File: rocketpy/motors/Tank.py [hunk 4]
```diff
@@ -0,0 +1,842 @@
...
+        -------
+        Function
+            Tank's liquid volume as a function of time.
+        """
+        pass
+
+    @functools.cached_property
+    def centerOfMass(self):
+        """Returns the center of mass of the tank's fluids as a function of
+        time.
```
Comment (giovaniceotto) [inline rocketpy/motors/Tank.py:234]: TODO: Specify relative to which point is the center of mass calculated (coordinate system orientation and origin).

File: rocketpy/motors/Tank.py [hunk 5]
```diff
@@ -0,0 +1,842 @@
...
+        self.cylinderGasMass = self.gas.density * self.cylinder.empty_volume
+        self.upperCapGasMass = self.gas.density * self.upperCap.empty_volume
+
+        self.bottomCapMass = self.bottomCapLiquidMass + self.bottomCapGasMass
+        self.cylinderMass = self.cylinderLiquidMass + self.cylinderGasMass
+        self.upperCapMass = self.upperCapLiquidMass + self.upperCapGasMass
+
+    def evaluateCentroids(self):
+        """Calculates the centroids of the liquid and gaseous portions of the tank
+        based on the volume distributions.
```
Comment (giovaniceotto) [inline rocketpy/motors/Tank.py:126]: TODO: Make it clear relative to which point are the centroids calculated (coordinate system orientation and origin).

File: rocketpy/motors/Tank.py [hunk 6]
```diff
@@ -0,0 +1,842 @@
...
+        centerOfMass = Function(
+            centerOfMass, inputs="Time (s)", outputs="Center of Mass (m)"
+        )
+
+        return centerOfMass
+
+    @functools.cached_property
+    def inertiaTensor(self):
+        """Returns the inertia tensor of the tank's fluids as a function of
+        time.
```
Comment (giovaniceotto) [inline rocketpy/motors/Tank.py:278]: TODO: make it clear relative to which origin and axes (coordinate system) is this being calculated.

File: rocketpy/motors/Tank.py [hunk 7]
```diff
@@ -0,0 +1,842 @@
...
+            Center of mass of the tank's fluids as a function of time.
+        """
+
+        def centerOfMass(t):
+            self.evaluateTankState(t)
+
+            bottomCapMassBalance = (
+                self.bottomCapLiquidMass * self.bottomCap.filled_centroid
+                + self.bottomCapGasMass * self.bottomCap.empty_centroid
+            )
```
Comment (giovaniceotto) [inline rocketpy/motors/Tank.py:253]: No real change, just a suggestion to make things more symmetric.

```suggestion
            bottomCapMassBalance = (
                self.bottomCapLiquidMass * self.bottomCapLiquidCentroid
                + self.bottomCapGasMass * self.bottomCapGasCentroid
            )
```

File: rocketpy/motors/TankGeometry.py [hunk 8]
```diff
@@ -0,0 +1,464 @@
...
+            Height of the geometry, in meters. If not provided, is set as zero
+            for two-dimenstional geometries or equal to the radius for tri-dimensional.
+        filled_volume : float, optional
+            Volume of the geometry that is filled with propellant, in meters cubed.
+            If not provided, is set as zero.
+        fill_direction : str, optional
+            Direction of propellant filling. Relevant for tri-dimensional geometries
+            which are not vertically simetrical. Can be either "upwards" or "downwards",
+            as such:
+                - "upwards": propellant is filled with the convex site facing down, as in
```
Comment (giovaniceotto) [inline rocketpy/motors/TankGeometry.py:32]: > 
```suggestion
                - "upwards": propellant is filled with the convex side facing down, as in
```

File: rocketpy/motors/TankGeometry.py [hunk 9]
```diff
@@ -0,0 +1,464 @@
...
+                a cup of water. Most used for bottom caps.
+                - "downwards": propellant is filled with the concave site facing down, as
```
Comment (giovaniceotto) [inline rocketpy/motors/TankGeometry.py:34]:
```suggestion
                - "downwards": propellant is filled with the concave side facing down, as
```

File: rocketpy/motors/TankGeometry.py [hunk 10]
```diff
@@ -0,0 +1,464 @@
...
+                in an inverted cup. Most used for upper caps.
+            The default is "upwards", and will be used if the user commits a typing error.
```
Comment (giovaniceotto) [inline rocketpy/motors/TankGeometry.py:36]: Seems dangerous to do this. If the user writes "down", it is use "upwards". I believe implementing error handling would be more appropriate here.

**Review by giovaniceotto — CHANGES_REQUESTED** [review]
Almost there! Implementation is looking awesome. I just made a small number of suggestions.

**Comment by phmbressan** [comment]
Thank you for the review, great points were made! I will analyse and implement the suggested changes today.

**Review by giovaniceotto — APPROVED** [review]

**Pull request merged**